src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi

```diff
@@ -0,0 +1,248 @@
+    cdef bytes request_raw = receive_ops[0].message()
+    cdef object request_message
+    if method_handler.request_deserializer:
+        request_message = method_handler.request_deserializer(request_raw)
+    else:
+        request_message = request_raw
+
+    # Executes application logic
+    cdef object response_message = await method_handler.unary_unary(request_message, _ServicerContextPlaceHolder())
+
+    # Serializes the response message
+    cdef bytes response_raw
+    if method_handler.response_serializer:
+        response_raw = method_handler.response_serializer(response_message)
+    else:
+        response_raw = response_message
+
+    # Sends response message
+    cdef tuple send_ops = (
+        SendStatusFromServerOperation(
+        tuple(), StatusCode.ok, b'', _EMPTY_FLAGS),
+        SendInitialMetadataOperation(tuple(), _EMPTY_FLAGS),
+        SendMessageOperation(response_raw, _EMPTY_FLAGS),
+    )
+    await callback_start_batch(rpc_state, send_ops, loop)
+
+
+async def _handle_rpc(_AioServerState server_state, RPCState rpc_state, object loop):
+    # Finds the method handler (application logic)
+    cdef object method_handler = _find_method_handler(
+        rpc_state,
+        server_state.generic_handlers
+    )
+    if method_handler is None:
+        # TODO(lidiz) return unimplemented error to client side
+        raise NotImplementedError()
+    # TODO(lidiz) extend to all 4 types of RPC
+    if method_handler.request_streaming or method_handler.response_streaming:
+        raise NotImplementedError()
+    else:
+        await _handle_unary_unary_rpc(
+            method_handler,
+            rpc_state,
+            loop
+        )
+
+
+async def _server_call_request_call(_AioServerState server_state, object loop):
+    cdef grpc_call_error error
+    cdef RPCState rpc_state = RPCState()
+    cdef object future = loop.create_future()
+    cdef CallbackWrapper wrapper = CallbackWrapper(future)
+    # NOTE(lidiz) Without Py_INCREF, the wrapper object will be destructed
+    # when calling "await". This is an over-optimization by Cython.
+    cpython.Py_INCREF(wrapper)
+    error = grpc_server_request_call(
+        server_state.server.c_server, &rpc_state.call, &rpc_state.details,
+        &rpc_state.request_metadata,
+        server_state.cq, server_state.cq,
+        wrapper.c_functor()
+    )
+    if error != GRPC_CALL_OK:
+        raise RuntimeError("Error in _server_call_request_call: %s" % error)
+
+    await future
+    cpython.Py_DECREF(wrapper)
+    return rpc_state
+
+
+async def _server_main_loop(_AioServerState server_state):
+    cdef object loop = asyncio.get_event_loop()
+    cdef RPCState rpc_state
+    cdef object waiter
+
+    while True:
+        rpc_state = await _server_call_request_call(
+            server_state,
+            loop)
+        # await waiter
+
+        loop.create_task(_handle_rpc(server_state, rpc_state, loop))
+        await asyncio.sleep(0)
```

Comment:
The coroutine will break the execution returning back the control to the loop here [1], no matter if gRPC already has something for being delivered - like a pending request, the `future` that is being awaited within the `_server_call_request_call` function will have always to yield the execution for at least one loop iteration which would give enough time for executing the already accepted request.,

The only reason that I would think that this could help will be an optimization for giving more priority to the current request. But TBH doing this, we will start polluting our code with `await asyncio.sleep(0)` when we do not have yet probes that this is really helping a lot.


[1] https://github.com/grpc/grpc/pull/20598/files#diff-b53c077f7911f53dc8ce7656e16d35cdR184

Reply:
I have to say, I was wrong about the hanging. I did benchmark with and without `asyncio.sleep(0)`. Turn out, the without version performs better in short bursts. In longer scenario, the QPS is equivalent to the original version. The per RPC latency is not impacted by this change.

```
lidiz@ddev:grpc-go-client$ go run benchmarker.go
Finished 40000 RPCs in 2.197975732s, QPS 18198.56
Through put is 17.8 KiB per second
lidiz@ddev:grpc-go-client$ go run benchmarker.go -t 1000000
Finished 1000000 RPCs in 1m20.167130808s, QPS 12473.94
Through put is 12.2 KiB per second
```

The `asyncio.sleep(0)` is removed.